test/hmactest.c

```diff
@@ -79,224 +82,176 @@ static struct test_st {
     ebcdic2ascii(test[2].key, test[2].key, test[2].key_len);
     ebcdic2ascii(test[2].data, test[2].data, test[2].data_len);
 #  endif
 
-    for (i = 0; i < 4; i++) {
-        p = pt(HMAC(EVP_md5(),
-                    test[i].key, test[i].key_len,
-                    test[i].data, test[i].data_len, NULL, NULL),
-                    MD5_DIGEST_LENGTH);
-
-        if (strcmp(p, (char *)test[i].digest) != 0) {
-            printf("Error calculating HMAC on %d entry'\n", i);
-            printf("got %s instead of %s\n", p, test[i].digest);
-            err++;
-        } else
-            printf("test %d ok\n", i);
-    }
-# endif                         /* OPENSSL_NO_MD5 */
+    p = pt(HMAC(EVP_md5(),
+                test[idx].key, test[idx].key_len,
+                test[idx].data, test[idx].data_len, NULL, NULL),
+                MD5_DIGEST_LENGTH);
+
+    if (!TEST_str_eq(p, (char *)test[idx].digest))
+      return 0;
+
+    return 1;
+}
+# endif
+
+static int test_hmac_bad()
+{
+    HMAC_CTX *ctx = NULL;
+    int ret = 0;
 
-/* test4 */
     ctx = HMAC_CTX_new();
-    if (ctx == NULL) {
-        printf("HMAC malloc failure (test 4)\n");
-        err++;
-        goto end;
-    }
-    if (HMAC_CTX_get_md(ctx) != NULL) {
-        printf("Message digest not NULL for HMAC (test 4)\n");
-        err++;
-        goto test5;
-    }
-    if (HMAC_Init_ex(ctx, NULL, 0, NULL, NULL)) {
-        printf("Should fail to initialise HMAC with empty MD and key (test 4)\n");
-        err++;
-        goto test5;
-    }
-    if (HMAC_Update(ctx, test[4].data, test[4].data_len)) {
-        printf("Should fail HMAC_Update with ctx not set up (test 4)\n");
-        err++;
-        goto test5;
-    }
-    if (HMAC_Init_ex(ctx, NULL, 0, EVP_sha1(), NULL)) {
-        printf("Should fail to initialise HMAC with empty key (test 4)\n");
-        err++;
-        goto test5;
-    }
-    if (HMAC_Update(ctx, test[4].data, test[4].data_len)) {
-        printf("Should fail HMAC_Update with ctx not set up (test 4)\n");
-        err++;
-        goto test5;
-    }
-    printf("test 4 ok\n");
-test5:
-    /* Test 5 has empty key; test that single-shot accepts a NULL key. */
-    p = pt(HMAC(EVP_sha1(), NULL, 0, test[4].data, test[4].data_len,
-                NULL, NULL), SHA_DIGEST_LENGTH);
-    if (strcmp(p, (char *)test[4].digest) != 0) {
-        printf("Error calculating HMAC on %d entry'\n", i);
-        printf("got %s instead of %s\n", p, test[4].digest);
-        err++;
-    }
+    if (!TEST_ptr(ctx))
+        goto err;
 
+    if (!TEST_ptr_null(HMAC_CTX_get_md(ctx)))
+        goto err;
+
+    if (!TEST_false(HMAC_Init_ex(ctx, NULL, 0, NULL, NULL)))
+        goto err;
+
+    if (!TEST_false(HMAC_Update(ctx, test[4].data, test[4].data_len)))
+        goto err;
+
+    if (!TEST_false(HMAC_Init_ex(ctx, NULL, 0, EVP_sha1(), NULL)))
+        goto err;
+
+    if (!TEST_false(HMAC_Update(ctx, test[4].data, test[4].data_len)))
```

Comment:
I'd prefer this in a single if - more consistent with our style elsewhere. Similar comment for other places in this file.